### PR TITLE
WT-6624 Use transaction snapshot for applications performing eviction

### PR DESCRIPTION
**Pull request by hbokhari**
If we have entered eviction through application threads and we are running transaction with snapshot isolation, we want to update our snapshot to evict pages that are not globally visible based on `last_running` transaction. We will save some of the running transaction's context, perform eviction and restore the original transaction's context (including snapshot) once we are finished.

### CODE REVIEW

**Comment by hbokhari** [comment]
@bvpvamsikrishna This is something I think will be very close to a final solution. More changes will come in once WT-6490 get merged.

File: src/evict/evict_page.c
Comment (sulabhM) [inline src/evict/evict_page.c:699]: "use_snapshot_for_app_thread" and "use_snapshot_for_eviction_thread" naming convention signals to me that only in certain cases we use snapshot with application thread and with eviction threads. If my understanding is correct, we also use a snapshot with the application thread unless the metadata is being updated and we always take a snapshot with the eviction thread.

Maybe we can merge the two booleans together into one "acquire_updated_snapshot"

File: src/evict/evict_page.c
Comment (sulabhM) [inline src/evict/evict_page.c:716]: possibly merge the two bools as mentioned above and use "! FLD_ISSET(evict_flags, WT_REC_EVICTION_THREAD);" here ?

File: src/evict/evict_page.c
Comment (sulabhM) [inline src/evict/evict_page.c:731]: I liked the "snapshot_acquired" name better. It signals what has happened rather than what we need to do. Feel free to ignore me on this one if you disagree.

File: src/txn/txn.c
Comment (sulabhM) [inline src/txn/txn.c:287]: I do not understand what we are using snapshot_buffer for and why it needs to be "sizeof(txn->snapshot[0]) * S2C(session)->session_size" large. It is likely my ignorance about the snapshot management and transaction code.

**Review by sulabhM — CHANGES_REQUESTED** [review]

File: src/txn/txn.c
Comment (hbokhari) [reply src/txn/txn.c:287]: We want to acquire a new snapshot while reserving the existing snapshot. `snapshot_buffer` is temporary place holder to allocate a new array and assign it `txn::snapshot`. The array should be large enough to hold the number of sessions that can be active (`WT_CONN_IMPL::session_size`).

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c:694]: Fixed in d38f84e.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Agreed. Fixed in c05646f.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Yes it was an accidental change. Fixed in d767dd1.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Fixed in e11c52c.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c:698]: I used two boolean to make it easier to understand the conditions required for each scenario and to make the final if condition a bit simpler. If you believe if it's okay to combine the two then I will make the change.

File: src/reconcile/rec_visibility.c
Comment (agorrod) [inline src/reconcile/rec_visibility.c:289]: Would something similar to this be enough without copying out and allocating a new snapshot for each reconciliation?

i.e: the correctness problem with doing snapshot based reconciliation using an application thread is that it's possible to evict uncommitted content associated with the currently running transaction. Could we just use the existing snapshot and keep this check?

Comment (hbokhari) [reply src/reconcile/rec_visibility.c:289]: @agorrod - We can just add this check and force isolation level to improve the efficacy of eviction by application threads. However, it will not match the performance we can achieve by taking an updated snapshot. Essentially, we will not be able to evict committed changes belonging to transactions in snapshot or transaction ids higher than `snap_max`.

**Comment by tammybailey** [comment]
@hbokhari what do we need to move this one along? We'd like to get this fix in so we can address WT-6664.

**Comment by hbokhari** [comment]
@tammybailey I waiting for a review from @agorrod and @bvpvamsikrishna.

File: src/reconcile/rec_visibility.c
Comment (agorrod) [reply src/reconcile/rec_visibility.c:289]: I think that compromise is acceptable given the reduced complexity of the solution.
If we encounter cases where it's causing problems we can do something more similar to the complete solution here.

**Review by agorrod — CHANGES_REQUESTED** [review]

File: src/reconcile/rec_visibility.c
Comment (hbokhari) [reply src/reconcile/rec_visibility.c:289]: @agorrod I have pushed 38dc1cb where we are use existing snapshot for application threads.

**Review by sulabhM — APPROVED** [review]

**Review by agorrod — APPROVED** [review]
Thanks @hbokhari this looks better to me now.
Please run a full stress test suite and a MongoDB test suite before merging, as I think this is a risky change.

I'm also going to add @sueloverso as a reviewer - I haven't done a detailed style review here.

File: src/evict/evict_page.c
Comment (sueloverso) [inline src/evict/evict_page.c:686]: ...Otherwise it becomes harder in the later part...
...to detect updates that belonged...

File: src/evict/evict_page.c
Comment (sueloverso) [reply src/evict/evict_page.c:694]: I would prefer one boolean. By having two and the complex conditions here, it seems it would be bad if both were true or both were false. Please combine if you can.

File: src/evict/evict_page.c
Comment (sueloverso) [inline src/evict/evict_page.c:549]: We are in eviction already, so I'm not finding the extra words here additionally helpful.

File: src/evict/evict_page.c
Comment (sueloverso) [inline src/evict/evict_page.c:681]: Text fix: ...if coming through the eviction thread...

File: src/evict/evict_page.c
Comment (sueloverso) [inline src/evict/evict_page.c:683]: ...based on the last_running transaction...

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Fixed in 65be4b5.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Fixed in 65be4b5.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c]: Fixed in 65be4b5.

File: src/evict/evict_page.c
Comment (hbokhari) [reply src/evict/evict_page.c:695]: Each scenario (application and eviction thread) has a different criteria for using snapshot for eviction. I have added an assert to make sure either only one or none of the booleans is set to true (67b7032),

File: src/evict/evict_page.c
Comment (sueloverso) [inline src/evict/evict_page.c:698]: The simplicity of that boolean makes me think it really isn't necessary. I'd still prefer only one. If you really want to keep it, can it be called `is_eviction_thread` instead? That more accurately describes what it is.

**Review by sueloverso — APPROVED** [review]
A few minor style things.

**Pull request merged**
